Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+where
+  /-- Given a `pf` of an `Iff`, `Eq`, or `HEq`, return a proof of `Eq`.
+  If `pf` is not obviously any of these, weakly try inserting `propext` to make an `Iff`
+  and otherwise unify the type with `Eq`. -/
+  toEqPf : MetaM Expr := do
+    let ty ← whnf (← inferType pf)
+    if let some .. := ty.iff? then
+      mkPropExt pf
+    else if let some .. := ty.eq? then
+      return pf
+    else if let some (lhsTy, _, rhsTy, _) := ty.heq? then
+      unless ← isDefEq lhsTy rhsTy do
+        throwError "Cannot turn HEq proof into an equality proof. Has type{indentD ty}"
+      mkAppM ``eq_of_heq #[pf]
+    else if ← Meta.isProp lhs then
+      mkPropExt (← ensureIff pf)
+    else
+      discard <| mkEqForExpectedType (← inferType pf)
+      return pf
+  /-- Given a `pf` of an `Iff`, `Eq`, or `HEq`, return a proof of `HEq`.
+  If `pf` is not obviously any of these, weakly try making it be an `Eq` or an `Iff`,
+  and otherwise make it be a `HEq`. -/
+  toHEqPf : MetaM Expr := do
+    let ty ← whnf (← inferType pf)
+    if let some .. := ty.iff? then
+      mkAppM ``heq_of_eq #[← mkPropExt pf]
+    else if let some .. := ty.eq? then
+      mkAppM ``heq_of_eq #[pf]
+    else if let some .. := ty.heq? then
+      return pf
+    else if ← withNewMCtxDepth <| isDefEq (← inferType lhs) (← inferType rhs) then
+      mkAppM ``heq_of_eq #[← toEqPf]
+    else
+      discard <| mkHEqForExpectedType (← inferType pf)
+      return pf
+  ensureSidesDefeq (pf : Expr) : MetaM Expr := do
+    let pfTy ← inferType pf
+    let some (_, lhs', _, rhs') ← sides? pfTy
+      | panic! "Unexpectedly did not generate an eq or heq"
+    unless ← isDefEq lhs lhs' do
+      throwError "Congruence hole has type{indentD pfTy}\n{""
+        }but its left-hand side is not definitionally equal to the expected value{indentD lhs}"
+    unless ← isDefEq rhs rhs' do
+      throwError "Congruence hole has type{indentD pfTy}\n{""
+        }but its right-hand side is not definitionally equal to the expected value{indentD rhs}"
+    return pf
+
+/-- Force the lhs and rhs to be defeq. For when `dsimp`-like congruence is necessary.
+Clears the proof. -/
+def CongrResult.defeq (res : CongrResult) : MetaM CongrResult := do
+  if res.isRfl then
+    return res
+  else
+    unless ← isDefEq res.lhs res.rhs do
+      throwError "Cannot generate congruence because we need{indentD res.lhs}\n{""
+        }to be definitionally equal to{indentD res.rhs}"
+    -- Propagate types into any proofs that we're dropping:
+    discard <| res.eq
+    return {res with pf? := none}
+
+/-- Tries to make a congruence between `lhs` and `rhs` automatically.
+1. If they are defeq, returns a trivial congruence.
+2. Tries using `Subsingleton.elim`.
+3. Tries `proof_irrel_heq` as another effort to avoid doing congruence on proofs.
+3. Otherwise throws an error.
+
+Note: `mkAppM` uses `withNewMCtxDepth`, which prevents `Prop` specialization.
+Otherwise the `Subsingleton Prop` instance could apply for example. -/
```

Comment:
I'm confused by this note.

Reply:
This was a problem in `congr!` as well. I added some clarification.

The issue is that if you have `Sort ?u` then if there weren't `withNewMCtxDepth` then instance search is perfectly happy to unify this with `Prop` and say "yes, these are subsingletons". Without it, this would succeed:
```
example (T : Sort _) (x y : T) : x = y := congr(_)
```